pycqed/measurement/waveform_control/segment.py

```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
                 el_new_start = el_list[i + 1][0]
 
                 if (el_prev_end - el_new_start) > tol :
-                    print(el_prev_end, el_new_start)
-                    msg = f'{prev_el} (ends at {el_prev_end*1e6:.4f}us) and ' \
-                    f'{el_list[i + 1][2]} (' \
-                        f'starts at {el_new_start*1e6:.4f}us) overlap ' \
-                          f'on {awg}'
-                    if allow_overlap:
-                        log.warning(msg)
+                    if track_and_ignore:
+                        # add set of two overlapping elements to list
+                        self.overlapping_elements.append({prev_el,
+                                                          el_list[i + 1][2]})
                     else:
-                        raise ValueError(msg)
+                        print(el_prev_end, el_new_start)
+                        msg = f'{prev_el} (ends at {el_prev_end*1e6:.4f}us) ' \
+                              f'and {el_list[i + 1][2]} (' \
+                              f'starts at {el_new_start*1e6:.4f}us) overlap ' \
+                              f'on {awg}'
+                        if allow_overlap:
+                            log.warning(msg)
+                        else:
+                            raise ValueError(msg)
+
+    def resolve_overlap(self):
+        """
+        Routine to resolve overlapping elements. Will be exectued if
+        corresponding pulsar parameter is true. This code
+        first goes through the list of overlapping elements that are
+        pairwise overlapping and clusters them into lists
+        of overlapping elements, where two different lists of overlapping
+        elements have no overlapping elements with
+        one another. At the end the code combines all elements of each
+        list into a new element.
+        """
+
+        self.gen_elements_on_awg()
+        self._test_overlap(track_and_ignore=True)
+        overlapping_elements = self.overlapping_elements
+
+        if len(overlapping_elements)==0:
+            return
+
+        # add first two overlapping elements to list
+        joint_overlapping_elements = [overlapping_elements[0]]
+
+        new_cluster = True
+        for i in range(len(overlapping_elements) - 1):
+            # check for all sets added to joint_overlapping_elements
+            # whether the next set of elements from
+            # overlapping_elements shares an element name. If so,
+            # then add this element to the entry in
+            # joint_overlapping_elements
+            for j in range(len(joint_overlapping_elements)):
```

Comment:
Unless I am missing something, the check via this loop is more complicated than needed. overlapping_elements should be in correct order due to `el_list.sort()` in `_test_overlap`, so is it really needed to check all entries of `joint_overlapping_elements` here?

Reply:
I think you are right, that it is sufficient to only test whether it it in the last entry of `joint_overlapping_elements`. I will test it like this, even though it is always difficult to come up with edge cases for which this might not work.

Reply:
Seems to work, nice catch!